### PR TITLE
Support conversion of local links to md files (disabled by default, c…

### LINKED ISSUE #50: Support automatic linking
I am currently converting a large documentation site from Confluence to docnet. So far so good, but I noticed (it's also documented) that links to local md files don't get transformed.

Interested in a PR?

If so, links will be converted if:

1. They are listed in docnet.json (so have a generated htm page)
2. Are relative

### PR DESCRIPTION
**Pull request by GeertvanHorrik**
Disabled by default, can be enabled by *ConvertLocalLinks="true"* in the config

Fixes #50

### CODE REVIEW

**Comment by GeertvanHorrik** [comment]
1 think left if it's really a strict requirement in LinkDefinition.cs:

// TODO: Check if link exists in the ToC

**Comment by FransBouma** [comment]
> 1 thing left if it's really a strict requirement in LinkDefinition.cs:

what do you mean by this exactly? 

Btw, could you add a test which handles `(LocalLink to Bla)[./somefile.md#bla]`

**Comment by GeertvanHorrik** [comment]
Will add the tests, should be supported though.

My idea was it only makes sense to convert md links when they are actually being converted to htm (thus included in the ToC). However, we're not aware of that context so deep in the parser / converter.

**Comment by FransBouma** [comment]
Ah makes sense, yes it's not clear to the parser which links are in scope, in fact that's a hard problem to do while not being finished with all the pages, I think. 

Thanks for the tests and code! Will try out the PR and merge if everything is in order.

File: src/MarkdownDeep/LinkDefinition.cs [hunk 1]
```diff
@@ -58,9 +58,36 @@ internal void RenderLink(Markdown m, StringBuilder b, string link_text, List<str
...
+			    {
+			        // A few requirements before we can convert local links:
+                    //   1. Link contains .md
+                    //   2. Link is relative
+                    //   3. Link is included in the index
+                    var index = url.LastIndexOf(".md", StringComparison.OrdinalIgnoreCase);
+			        if (index >= 0)
+			        {
+			            Uri uri;
+			            if (Uri.TryCreate(url, UriKind.RelativeOrAbsolute, out uri))
```
Comment (FransBouma) [inline src/MarkdownDeep/LinkDefinition.cs]:
```cs
if(Uri.TryCreate(url, UriKind.Relative, out uri))
```

will return true only if the uri is relative, which is what's required. Then the inner test isn't needed.

File: src/MarkdownDeep/LinkDefinition.cs [hunk 2]
```diff
@@ -58,9 +58,36 @@ internal void RenderLink(Markdown m, StringBuilder b, string link_text, List<str
...
+			            {
+			                if (!uri.IsAbsoluteUri)
```
Comment (FransBouma) [inline src/MarkdownDeep/LinkDefinition.cs]: Not needed if above uri creation is used.

File: src/MarkdownDeep/LinkDefinition.cs [hunk 3]
```diff
@@ -58,9 +58,36 @@ internal void RenderLink(Markdown m, StringBuilder b, string link_text, List<str
...
+			                {
+			                    // TODO: Check if link exists in the ToC
```
Comment (FransBouma) [inline src/MarkdownDeep/LinkDefinition.cs]: TODO comments are not really useful, I think. An issue is better :)

File: src/MarkdownDeep/LinkDefinition.cs [hunk 4]
```diff
@@ -58,9 +58,36 @@ internal void RenderLink(Markdown m, StringBuilder b, string link_text, List<str
...
+			                    var existsInTableOfContents = true;
+			                    if (existsInTableOfContents)
```
Comment (FransBouma) [inline src/MarkdownDeep/LinkDefinition.cs]: Not needed unless the feature to check for ToC presence is implemented. Can be removed.

File: src/MarkdownDeep/LinkDefinition.cs [hunk 5]
```diff
@@ -58,9 +58,36 @@ internal void RenderLink(Markdown m, StringBuilder b, string link_text, List<str
...
+			                    {
+			                        url = url.Remove(index, ".md".Length).Insert(index, ".htm");
```
Comment (FransBouma) [inline src/MarkdownDeep/LinkDefinition.cs]:
```cs
url = String.Concat(url.Substring(0, index), ".htm", url.Substring(index + ".md".Length));
```
is more efficient: .Remove() will create a new string with both ends concatenated together, then the insert will destroy that string again and allocate a new one. The String.Concat() will use a recycled stringbuilder from a pool to concat string together in 1 go. :) The ".md".Length can be hardcoded, as it's a constant, but for clarity I think this is better indeed.

File: src/MarkdownDeepTests/testfiles/locallinks_enabled/markdown-readme(DocNetMode)(ConvertLocalLinks).text [hunk 6]
```diff
@@ -0,0 +1,8 @@
+Markdown
+========
+
+[Remote link](https://github.com/FransBouma)
```
Comment (FransBouma) [inline src/MarkdownDeepTests/testfiles/locallinks_enabled/markdown-readme(DocNetMode)(ConvertLocalLinks).text]: A test is missing for: 
```
[Remote link](https://github.com/FransBouma/target1.md)
```
VERY minor, I know considering what the code does.

**Comment by FransBouma** [comment]
Oh and documentation... I'll add some docs about this feature too.

**Comment by FransBouma** [comment]
I'll release a new build in releases when your other PR is merged...

**Comment by FransBouma** [comment]
And now for the weird part... I can't get it to work with the docnet docs... Very strange... Debugging...

**Comment by FransBouma** [comment]
Scratch that, I'm an idiot. I formulated the URL markdown wrong!

**Comment by FransBouma** [comment]
FOr reference: http://fransbouma.github.io/DocNet/markdownsupport.htm#linking

**Comment by GeertvanHorrik** [comment]
Just checked the docs: I haven't tested the true, false values as strings, I just use it without quotes. Will prepare the other pr now.

**Comment by FransBouma** [comment]
oh I assumed I had to specify everything as strings, as it's json :) (it likely also converts "1" to true, but not sure). Didn't know you could specify `true` as a value without quotes!

**Pull request merged**